pyface/data_view/data_models/row_table_data_model.py

```diff
@@ -0,0 +1,126 @@
+# (C) Copyright 2005-2020 Enthought, Inc., Austin, TX
+# All rights reserved.
+#
+# This software is provided without warranty under the terms of the BSD
+# license included in LICENSE.txt and may be redistributed only under
+# the conditions described in the aforementioned license. The license
+# is also available online at http://www.enthought.com/licenses/BSD.txt
+#
+# Thanks for using Enthought open source!
+""" A row-oriented data model implementation.
+
+This module provides a concrete implementation of a data model for the
+case of row-oriented data.
+"""
+from traits.api import Any, ComparisonMode, Instance, List, observe
+
+from pyface.data_view.abstract_data_model import (
+    AbstractDataModel, DataViewSetError
+)
+from pyface.data_view.index_manager import IntIndexManager
+from pyface.data_view.data_models.data_accessors import AbstractDataAccessor
+
+
+class RowTableDataModel(AbstractDataModel):
+    """ A data model that presents a list of objects as rows.
+    """
+
+    #: A sequence of objects to display in columns.
+    data = Any()
```

Comment:
Related to the comment about observing mutation on a trait container: It is important that this is not a container trait, otherwise the data would be copied, rather than referenced. We want the container to be referenced. Perhaps we should add a comment as a reminder, because this is quite tempting otherwise to turn this trait into `Union(List, Set, Dict)` in the future.

Reply:
Actually, could we use `Instance(collections.abc.Sequence)` here? 🤔

Reply:
I haven't tested it, but I think a `DataFrame` out to be a valid thing to go here, for example (uncertainty is because I'm not sure if they've changed the indexing API in the last year or so).  Not sure if it is a `Sequence`.

Reply:
And for the record `isinstance(DataFrame, Sequence)` is false.

Reply:
I checked, pandas DataFrame is not a sequence, e.g. it does not implement `__reversed__`.
Looking at this model, looks like all this model needs are `__len__` and `__getitem__`.

Checking for `collections.abc.Sized` gives us `__len__`, and `DataFrame` does pass this test.
However for `__getitem__`, the next available thing is `collections.abc.Sequence`, which captures more than we need and DataFrame does not fulfil the check.

It might be worth documenting that `data` is required to support `__len__` and `__getitem__(int)`

Reply:
> And for the record isinstance(DataFrame, Sequence) is false.

You beat me to it :)

Reply:
I played around a bit with getting `DataFrame` working, and it really requires use of `iloc` rather than direct indexing, and given that `DataFrame` can be hierarchical that is pushing me towards `DataFrame` not being supported here.

So targets are `Sequence`-y sorts of things (although as you note, we don't need `__reversed__`) like lists, tuples, 1D or 2D arrays (and "array-like").  I can't think of anything which has `__len__` and `__getitem__` but not `__reversed__`, so may be `Instance(Sequence)` _is_ good enough?

Reply:
Yea, sounds good to me!

Reply:
Arg.  `Array` isn't a `Sequence` either.

Reply:
OK, let's do duck typing here and just specify what methods are required for this model then.
(Just found an interesting issue on numpy's repo about this, it is still an open issue.)

Reply:
For now, due to https://github.com/enthought/traits/issues/1277 I'm leaning towards `Instance(Sequence)` and we can expand as needed (or people can subclass).

Reply:
SGTM!